File: app/javascript/channels/device_channel.js
import consumer from "./consumer"

const deviceChannel = consumer.subscriptions.create("DeviceChannel", {
  connected() {
    console.log("Connected to device channel")
  },

  disconnected() {
    console.log("Disconnected from device channel")
  },

  received(data) {
    console.log("Received message:", data)
    
    if (window.Dashboard) {
      switch(data.type) {
        case 'sync_complete':
          window.Dashboard.handleSyncComplete?.(data)
          break
        case 'verification_request':
          window.Dashboard.handleVerificationRequest?.(data)
          break
        case 'device_update':
          window.Dashboard.handleDeviceUpdate?.(data)
          break
      }
    }
  }
})

export default deviceChannel

File: app/channels/device_channel.rb
class DeviceChannel < ApplicationCable::Channel
  def subscribed
    stream_from "device_channel_#{current_device.id}"
  end

  def unsubscribed
    stop_all_streams
  end

  def receive(data)
    case data['action']
    when 'sync_complete'
      broadcast_sync_complete(data)
    when 'device_update'
      broadcast_device_update(data)
    end
  end

  private

  def broadcast_sync_complete(data)
    ActionCable.server.broadcast("device_channel_#{current_device.id}", {
      type: 'sync_complete',
      status: data['status'],
      synced_at: Time.current
    })
  end

  def broadcast_device_update(data)
    ActionCable.server.broadcast("device_channel_#{current_device.id}", {
      type: 'device_update',
      device: data['device']
    })
  end
end
